mopidy_youtube/youtube.py

```diff
@@ -0,0 +1,413 @@
+            self._thumbnails = [
+                val['url']
+                for (key, val) in snip['thumbnails'].items()
+                if key in ['medium', 'high']
+            ]
+
+    def __init__(self, id):
+        super(Playlist, self).__init__(id)
+
+        self._title = None
+        self._thumbnails = None
+        self._video_count = None
+        self._videos = None
+        self._videos_event = None
+        self._info_event = None
+
+    def load_info(self):
+        self.load_info_mult([self])
+        return self._info_event
+
+    # loads list of videos of a playlist using one API call for every 50
+    # fetched videos. For every page fetched, Video.load_info_mult is called to
+    # start loading video info in a separate thread.
+    #
+    def load_videos(self):
+        if self._videos is not None or self._videos_event:
+            return self._videos_event
+
+        def job():
+            all_videos = []
+            page = ''
+            while page is not None and len(all_videos) < self.max_videos:
+                query = {
+                    'part': 'id,snippet',
+                    'fields': 'nextPageToken,' +
+                              'items(snippet(title,resourceId(videoId)))',
+                    'maxResults': min(self.max_videos - len(all_videos), 50),
+                    'playlistId': self.id,
+                    'key': yt_key,
+                    'pageToken': page,
+                }
+                result = session.get(yt_api_endpoint+'playlistItems',
+                                     params=query)
+                data = result.json()
+                page = data.get('nextPageToken') or None
+
+                videos = []
+                for item in data['items']:
+                    video = Video.get(item['snippet']['resourceId']['videoId'])
+                    video._load_api_data(item)
+                    videos.append(video)
+                all_videos += videos
+
+                # start loading video info for this batch in the background
+                Video.load_info_mult(videos)
+
+            self._videos = all_videos
+
+        self._videos_event = ThreadPool.run(job)
+        return self._videos_event
+
+    @property
+    def is_video(self):
+        return False
+
+
+# create methods for fetching properties loaded asynchronously
+# eg video.pafy, playlist.videos
+#
+Entry.make_async_prop('title', 'load_info')
+Entry.make_async_prop('channel', 'load_info')
+
+Video.make_async_prop('pafy', 'load_pafy')
+Video.make_async_prop('length', 'load_info')
+
+Playlist.make_async_prop('videos', 'load_videos')
+Playlist.make_async_prop('thumbnails', 'load_info')
+Playlist.make_async_prop('video_count', 'load_info')
+
+
+# simple 'dynamic' thread pool. Threads are created when new jobs arrive, stay
+# active for as long as there are active jobs, and get destroyed afterwards
+# (so that there are no long-term threads staying active)
+#
+class ThreadPool:
```

Comment:
I now use pykka.ThreadingFuture for async properties, makes things much cleaner.

On the other hand I couldn't figure out how to use pykka's actors for what I want. My constraints:
* a single job should be able to update the info for multiple videos/playlists
* no more than 15 threads should be running (shouldn't flood the network with queries when the user adds 200 videos to the playing queue)
* threads should run as long as there are jobs to do, then stop (no need to keep resources forever)

A simple thread pool, together with ThreadingFuture's does the job fine. Let me know whether it's ok or if you have suggestions about how to use actors (I have no experience with pykka).